argoverse/evaluation/stereo/utils.py

```diff
@@ -0,0 +1,252 @@
+    gt_obj_disparity = np.float32(gt_obj_disparity) / DISPARITY_NORMALIZATION
+
+    errors = accumulate_stereo_metrics(abs_error_thresholds)
+
+    # Compute masks
+    pred_mask = pred_disparity > 0
+    gt_mask = gt_disparity > 0
+    obj_mask = gt_obj_disparity > 0
+    bg_mask = np.logical_and(gt_mask, np.logical_not(obj_mask))
+    fg_mask = np.logical_and(gt_mask, obj_mask)
+
+    # If the density of the predicted disparity is less than 100%, then interpolate to fill up the holes.
+    num_pixels_all = pred_disparity.size
+    num_pixels_all_est = np.sum(pred_mask)
+    density = num_pixels_all_est / max(num_pixels_all, 1.0)
+
+    if density < 1.0:
+        pred_disparity = interpolate_disparity(pred_disparity)
+
+    # Compute errors
+    abs_err = np.abs(pred_disparity - gt_disparity)
+    rel_err = abs_err / np.maximum(gt_disparity, 1)
+
+    errors["num_pixels_bg"] = np.sum(bg_mask)
+    errors["num_pixels_fg"] = np.sum(fg_mask)
+    errors["num_pixels_bg_est"] = np.sum(bg_mask & pred_mask)
+    errors["num_pixels_fg_est"] = np.sum(fg_mask & pred_mask)
+
+    for abs_error_thresh, rel_error_thresh in zip(abs_error_thresholds, rel_error_thresholds):
+        bad_pixels = (abs_err > abs_error_thresh) & (rel_err > rel_error_thresh)
+
+        errors[f"num_errors_bg:{abs_error_thresh}"] = np.sum(bg_mask & bad_pixels)
+        errors[f"num_errors_fg:{abs_error_thresh}"] = np.sum(fg_mask & bad_pixels)
+        errors[f"num_errors_bg_est:{abs_error_thresh}"] = np.sum(bg_mask & pred_mask & bad_pixels)
+        errors[f"num_errors_fg_est:{abs_error_thresh}"] = np.sum(fg_mask & pred_mask & bad_pixels)
+
+    if save_disparity_error_image:
+        timestamp = int(Path(gt_fpath).stem.split("_")[-1])
+        log_id = Path(gt_fpath).parts[-3]
+        write_dir_path = Path(f"{figs_fpath}/{log_id}")
+
+        write_disparity_error_image(
+            pred_disparity,
+            gt_disparity,
+            timestamp,
+            write_dir_path,
+            abs_error_thresholds,
+            rel_error_thresholds,
+        )
+
+    return errors
+
+
+def accumulate_stereo_metrics(abs_error_thresholds: List[int]) -> pd.DataFrame:
+    """Stereo metrics accumulator.
+    Initializes the stereo metrics accumulator with zeroes in all fields.
+    """
+    num_fields = NUM_EVAL_PIXEL_REGIONS + NUM_EVAL_ERROR_REGIONS * len(abs_error_thresholds)
+
+    columns = [
+        "num_pixels_bg",
+        "num_pixels_fg",
+        "num_pixels_bg_est",
+        "num_pixels_fg_est",
+    ]
+
+    for abs_error_thresh in abs_error_thresholds:
+        columns += [
+            f"num_errors_bg:{abs_error_thresh}",
+            f"num_errors_fg:{abs_error_thresh}",
+            f"num_errors_bg_est:{abs_error_thresh}",
+            f"num_errors_fg_est:{abs_error_thresh}",
+        ]
+
+    return pd.DataFrame([[0] * num_fields], columns=columns)
+
+
+def interpolate_disparity(disp: np.ndarray) -> np.ndarray:
+    """Interpolate disparity image to inpaint holes.
```

Comment:
might be helpful to add 5-10 words just explaining is it "nearest neighbor" interpolation, or "linear", or "bilinear"? is there a URL or reference we could add to the docstring detailing the algorithm?

Reply:
sure